pyuvdata/utils.py

```diff
@@ -1518,27 +1529,260 @@ def uvcalibrate(
+                        "with calibration and flag the data for this antenna. ",
+                        DeprecationWarning,
+                    )
+                else:
+                    warn_str = (
+                        f"Antenna {this_ant_name} has data on UVData but not on UVCal. "
+                        "Data for this antenna will not be calibrated"
+                    )
+                    if flag_missing:
+                        warn_str += (
+                            " and will be flagged because flag_missing is True. "
+                            "The flag_missing keyword will be deprected in "
+                            "version 2.2, use override_ant_check instead."
+                        )
+                    else:
+                        warn_str += (
+                            ", set override_ant_check=True to proceed "
+                            "with calibration and flag the data for this antenna. "
+                            "Currently calibration will proceed, the data will not "
+                            "be flagged and not be calibrated. This behavior will "
+                            "deprected in version 2.2."
+                        )
+                    warnings.warn(warn_str, DeprecationWarning)
+
+    uvdata_times = np.unique(uvdata.time_array)
+    downselect_cal_times = False
+    if uvcal.Ntimes > 1:
+        uvcal_times = np.unique(uvcal.time_array)
+        try:
+            time_arr_match = np.allclose(
+                uvcal_times,
+                uvdata_times,
+                atol=uvdata._time_array.tols[1],
+                rtol=uvdata._time_array.tols[0],
+            )
+        except ValueError:
+            time_arr_match = False
+
+        if not time_arr_match:
+            # check more carefully
+            uvcal_times_to_keep = []
+            for this_time in uvdata_times:
+                wh_time_match = np.nonzero(
+                    np.isclose(
+                        uvcal.time_array - this_time,
+                        0,
+                        atol=uvdata._time_array.tols[1],
+                        rtol=uvdata._time_array.tols[0],
+                    )
+                )
+                if wh_time_match[0].size > 0:
+                    uvcal_times_to_keep.append(uvcal.time_array[wh_time_match][0])
+                else:
+                    warnings.warn(
+                        f"Time {this_time} exists on UVData but not on UVCal. "
+                        "This will become an error in version 2.2",
+                        DeprecationWarning,
+                    )
+            if len(uvcal_times_to_keep) < uvcal.Ntimes:
+                downselect_cal_times = True
+    elif uvcal.time_range is None:
+        if uvdata.Ntimes > 1 or not np.isclose(
+            uvdata_times,
+            uvcal.time_array,
+            atol=uvdata._time_array.tols[1],
+            rtol=uvdata._time_array.tols[0],
+        ):
```

Comment:
So this block is only entered in uvcal.Ntimes = 1 and uvcal.time_range is None. In that case, if UVData has multiple times the times cannot match. If it only has one time, you only need an isclose (rather than an allclose) because it's only testing one time. And you don't need it in a try/except because it won't have an exception (the exception happens with the allclose if the uvcal time array is shorter than the update time array I think). I added some inline comments to try to clarify, see if they're enough.